### ISSUE TITLE
plugin gives an error when File Explorer view is not activated at launch

### ISSUE THREAD

**Issue by ksdavidc** [alberti42/obsidian-import-attachments-plus#13]
1. If I install, the plugin is present, but not active. menu command to delete file is not present.
2. If I try ot activate it I get a message saying it failed to load.
3. If I go to the plugin info page I can load the plugin in, it shows as being active, and see the settings, but I get an error message. menu command to delete file is not present.
4. However, if i reload the vault, the plugin is inactive again!

[import attachmens.log](https://github.com/user-attachments/files/17748942/import.attachmens.log)

I am sure it is a plugin conflict but I have too many to troubleshoot right now. If you have any clue that might help me find the culprit that would be nice.

SYSTEM INFO:
	Obsidian version: v1.7.6
	Installer version: v1.7.5
	Operating system: Darwin Kernel Version 23.6.0: Mon Jul 29 21:13:00 PDT 2024; root:xnu-10063.141.2~1/RELEASE_X86_64 23.6.0
	Login status: not logged in
	Language: en
	Insider build toggle: off
	Live preview: on
	Base theme: dark
	Community theme: AnuPpuccin v1.5.0
	Snippets enabled: 13
	Restricted mode: off
	Plugins installed: 272
	Plugins enabled: 134
		1: Callout Suggestions v1.0.2
		2: Admonition v10.3.2
		3: Advanced Tables v0.22.1
		4: Another Quick Switcher v12.1.0
		5: Automatic Table Of Contents v1.4.0
		6: Better Command Palette v0.17.1
		7: Better Export PDF v1.9.2
		8: Better footnote v1.0.1
		9: BRAT v1.0.5
		10: CardBoard v0.7.9
		11: Change Case v2.0.128
		12: Doubleshift v2.2.1
		13: Colored Tags Wrangler v0.19.3
		14: Commander v0.5.2
		15: Convert url to preview (iframe) v0.5.0
		16: Copy document as HTML v0.8.1
		17: Crumbs v0.1.2
		18: Dataview v0.5.67
		19: Easy toggle sidebars v2.0.1
		20: Advanced URI v1.44.2
		21: Copilot v2.7.2
		22: Advanced Paste v2.7.0
		23: Control Characters v0.3.0
		24: Zoom v1.1.2
		25: Editing Toolbar v2.4.16
		26: Enhanced tables v1.0.5
		27: Export Image plugin v2.3.13
		28: Extract url content v0.12.1
		29: File Tree Alternative v2.6.0
		30: Find & Replace in Selection v1.1.0
		31: Folder Note Core v1.3.5
		32: Footnote Shortcut v0.1.3
		33: Get Info v1.1.0
		34: HelpMate v1.0.11
		35: Hide Sidebars on Window Resize v1.1.0
		36: Hover External Link Plugin v1.0.0
		37: Hyphenation v0.1.0
		38: Image in Editor v2.2.6
		39: Key-Value List v1.7.0
		40: Lapel v0.1.5
		41: Links v1.17.25
		42: Linter v1.27.1
		43: Local File Interface v0.1.0
		44: Local Images Plus v0.15.9
		45: Local REST API v2.5.4
		46: Merge Notes v0.0.9
		47: Markdown Table Editor v0.3.1
		48: Style Settings v1.0.9
		49: Open with v0.3.2
		50: Translate v1.4.9
		51: Toggle Case v1.1.0
		52: Tidy Footnotes v0.1.2
		53: Theme Toggler v1.0.1
		54: Theme Picker v1.0.8
		55: Text Transporter v1.0.12
		56: Text Snippets v0.1.2
		57: Text Format v3.1.0
		58: Text Extractor v0.5.2
		59: Text expand v0.11.4
		60: Templater v2.9.1
		61: Tasks Calendar Wrapper v0.3.1
		62: Tasks v7.13.0
		63: Task Collector (TC) v1.1.5
		64: Tag Wrangler v0.6.1
		65: Tag Project v0.9.0
		66: Table of Contents v0.2.0
		67: Paste URL into selection v1.7.0
		68: PDF break page v1.0.0
		69: Snippetor v0.4.5
		70: Size History v1.0.0
		71: Show Current File Path v0.5.2
		72: Sheets Extended v1.2.10
		73: Share Note v0.8.17
		74: Scratchblocks v0.0.1
		75: Reveal Active File Button v2.0.3
		76: Reset Font Size v1.0.2
		77: Regex Find/Replace v1.2.0
		78: QuickAdd v1.11.3
		79: Quick Tagger v2.2.3
		80: Quick Switcher++ v4.6.2
		81: Quick Explorer v0.2.11
		82: Projects v1.17.4
		83: Plugin Update Tracker v1.6.0
		84: Plugin Groups v2.1.0
		85: Pluck v1.0.1
		86: Outliner.md v0.1.9
		87: floating toc v2.4.7
		88: Kanban v2.0.51
		89: Excalidraw v2.6.7
		90: Outline Converter v1.3.2
		91: Digital Garden v2.57.2
		92: Datepicker v0.3.24
		93: Natural Language Dates v0.6.2
		94: MySnippets v1.2.3
		95: Note Refactor v1.8.2
		96: Edit mode switch v0.0.4
		97: Tagged Documents Viewer v1.0.2
		98: Tag Summary v2.1.5
		99: Number Headings v1.16.0
		100: Slides Extended v2.1.3
		101: Plugins Annotations v1.7.8
		102: Scroll to Top v2.1.4
		103: Query Control v0.6.0
		104: Auto Embed v1.2.4
		105: Settings Management v1.0.5
		106: Plugin Manager v0.1.5
		107: Share my plugin list v0.3.3
		108: Sort & Permute lines v0.7.0
		109: Icon Shortcodes v0.9.7
		110: Pane Relief v0.5.6
		111: Emoji Toolbar v0.4.1
		112: Table Generator v1.4.1
		113: Hover Editor v0.11.21
		114: Settings Search v1.3.10
		115: Calendar v1.5.10
		116: CardNote v1.2.0
		117: Enhance YouTube Links v1.0.8
		118: PDF++ v0.40.13
		119: Better PDF 2 v2.0.3
		120: Hotkey Helper v0.3.20
		121: Strange New Worlds v2.1.5
		122: Nav Link Header v1.0.3
		123: Copy Section v1.1.2
		124: Replace All v1.0.5
		125: Editing Mode Hotkey v1.0.0
		126: Heading Shifter v1.6.1
		127: Various Complements v10.0.3
		128: Omnisearch v1.25.1
		129: CSV Table v1.2.0
		130: Markdown Attributes v1.2.2
		131: Find Orphaned Images v1.0.0
		132: Lemons Search v0.0.7
		133: Broken Links v1.2.2
		134: Import Attachments+ v1.5.11

RECOMMENDATIONS:
	Custom theme and snippets: for cosmetic issues, please first try updating your theme and disabling your snippets. If still not fixed, please try to make the issue happen in the Sandbox Vault or disable community theme and snippets.
	Community plugins: for bugs, please first try updating all your plugins to latest. If still not fixed, please try to make the issue happen in the Sandbox Vault or disable community plugins.

**Comment by alberti42** [comment]
Hi David, thanks for the report! Sorry it is not working. Let's try to find out why.

I am not sure I will be able to easily understand from the .log file where it breaks. Could you replace the file `main.js` in the folder `<YOUR_VAULT>/.obsidian/plugins/import-attachments-plus` with the one uploaded   
[here](https://github.com/user-attachments/files/17749325/main.js.zip)? I believe the main.js you have is in the "production" version and therefore javascript-minified, which makes it not easy to understand where it breaks when you open it. However, if you open the file I provided, the code is clear.

You can then run your test and reproduce the error. Once the error pops up, it would be great if you could click on the error itself from the Obsidian console. You should see a link. It should take you to the lines of code where the code breaks. It would be helpful for me to know what function and line the code breaks in. So if you could take a screenshot, that would be really helpful.

**Comment by ksdavidc** [comment]
Here is the new console log with the links.

 I am not sure which one you want me to follow.
 
 Some point to the lines in your main.js file, aka the "VM372..." entries.

But, there are several "links" here (e.g. to anonymous functions, of the form. `app:js:1`, but with the actual would-be "line number" missing,
Do you need me to collect these as well?

```
app.js:1 Plugin failure: import-attachments-plus TypeError: Cannot convert undefined or null to object
    at Function.entries (<anonymous>)
    at updateVisibilityAttachmentFolders (VM372 plugin:import-attachments-plus:1195:14)
    at patchFileExplorer (VM372 plugin:import-attachments-plus:1224:5)
    at eval (VM372 plugin:import-attachments-plus:1840:7)
    at t.onLayoutReady (app.js:1:2158231)
    at ImportAttachments7.onload (VM372 plugin:import-attachments-plus:1839:24)
(anonymous) @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
s @ app.js:1
app.js:1 Plugin failure: import-attachments-plus TypeError: Cannot convert undefined or null to object
    at Function.entries (<anonymous>)
    at updateVisibilityAttachmentFolders (VM437 plugin:import-attachments-plus:1195:14)
    at patchFileExplorer (VM437 plugin:import-attachments-plus:1224:5)
    at eval (VM437 plugin:import-attachments-plus:1840:7)
    at t.onLayoutReady (app.js:1:2158231)
    at ImportAttachments7.onload (VM437 plugin:import-attachments-plus:1839:24)
(anonymous) @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
s @ app.js:1
Promise.then
l @ app.js:1
(anonymous) @ app.js:1
g @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
g @ app.js:1
e.enablePluginAndSave @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
g @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
t.onClick @ app.js:1
app.js:1 Plugin failure: import-attachments-plus TypeError: Cannot convert undefined or null to object
    at Function.entries (<anonymous>)
    at updateVisibilityAttachmentFolders (VM437 plugin:import-attachments-plus:1195:14)
    at patchFileExplorer (VM437 plugin:import-attachments-plus:1224:5)
    at eval (VM437 plugin:import-attachments-plus:1840:7)
    at t.onLayoutReady (app.js:1:2158231)
    at ImportAttachments7.onload (VM437 plugin:import-attachments-plus:1839:24)
(anonymous) @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
s @ app.js:1
Promise.then
l @ app.js:1
(anonymous) @ app.js:1
g @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
g @ app.js:1
e.enablePluginAndSave @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
g @ app.js:1
(anonymous) @ app.js:1
(anonymous) @ app.js:1
t.onClick @ app.js:1

```




The first app:js:1 goes to:

```
            e.prototype.enablePlugin = function(e, t) {
                return void 0 === t && (t = !1),
                g(this, void 0, Promise, (function() {
                    var n, i;
                    return y(this, (function(r) {
                        switch (r.label) {
                        case 0:
                            if (!(n = this.manifests[e]))
                                return [2, !1];
                            if (this.isDeprecated(n))
                                return new HL("Unable to load plugin ".concat(n.name, " v").concat(n.version, ". This version has been reported to cause issues. Please check for a newer version of the plugin.")),
                                [2, !1];
                            if ("sliding-panes-obsidian" === n.id)
                                return new HL(createFragment((function(e) {
                                    e.createEl("p", {
                                        text: "As of Obsidian v1.0, Sliding Panes is a built-in feature called Stacked Tabs. We recommend uninstalling the Sliding Panes plugin."
                                    }),
                                    e.createEl("p").createEl("a", {
                                        cls: "mod-cta",
                                        text: "Learn more",
                                        href: "https://help.obsidian.md/User+interface/Stacked+tabs",
                                        attr: {
                                            target: "_blank"
                                        }
                                    })
                                }
                                )),6e3),
                                [2, !1];
                            if ("better-pdf-plugin" === n.id && "1.4.0" === n.version)
                                return new HL("Better PDF Plugin is no longer functional. We recommend uninstalling it.",6e3),
                                [2, !1];
                            if (!Pl.isDesktopApp && n.isDesktopOnly)
                                return [2, !1];
                            r.label = 1;
                        case 1:
                            return r.trys.push([1, 3, 4, 5]),
                            this.loadingPluginId = e,
                            [4, this.loadPlugin(e, t)];
                        case 2:
                            return r.sent(),
                            this.loadingPluginId = null,
                            [3, 5];
                        case 3:
                            return i = r.sent(),
                            this.loadingPluginId = null,
                            new HL(Ef.interface.msgFailedToLoadPlugin({
                                plugin: e
                            })),
                            console.error("Plugin failure: " + e, i),
                            [2, !1];
                        case 4:
                            return K_("communityPlugins.".concat(e)),
                            [7];
                        case 5:
                            return [2, !0]
                        }
                    }
                    ))
                }
                ))
            }
     
```

The last goes to:

```
            t.prototype.onClick = function() {
                this.disabled || this.setValue(!this.getValue())
            }
```

**Comment by ksdavidc** [comment]
<img width="638" alt="Screenshot 2024-11-15 at 6 37 01 AM" src="https://github.com/user-attachments/assets/81fdb683-5240-447a-90fb-e0f1e040932a">
<img width="650" alt="Screenshot 2024-11-15 at 6 36 51 AM" src="https://github.com/user-attachments/assets/4cd016d1-4d44-4082-8e4f-3b1bd54ea74b">
<img width="653" alt="Screenshot 2024-11-15 at 6 36 39 AM" src="https://github.com/user-attachments/assets/17ae6022-3548-41b5-89fa-d0a6917fba7e">
<img width="664" alt="Screenshot 2024-11-15 at 6 36 28 AM" src="https://github.com/user-attachments/assets/f733a26a-2a41-47bc-8df8-11d3c0771cdb">
<img width="698" alt="Screenshot 2024-11-15 at 6 36 08 AM" src="https://github.com/user-attachments/assets/f7cfa034-720c-41de-8f09-d9143ae57038">
<img width="672" alt="Screenshot 2024-11-15 at 6 35 54 AM" src="https://github.com/user-attachments/assets/07bf5109-7a14-4660-90b8-b91ec08f7187">
<img width="636" alt="Screenshot 2024-11-15 at 6 35 32 AM" src="https://github.com/user-attachments/assets/d371058e-1725-4271-a69e-6483e185694c">

**Comment by alberti42** [comment]
Hi David, thanks a lot. In general, in the console, the error happens in the functions displayed in the beginning. The console shows the "history" of what function called what function until the error occurred. The lowest functions are typically not so important and related to deep down the internal functioning of Obsidian.

I presume something happens in the function

```typescript
function updateVisibilityAttachmentFolders(plugin: ImportAttachments){
	const leaves = plugin.app.workspace.getLeavesOfType('file-explorer');
	const hide = plugin.settings.hideAttachmentFolders;
	for (const leaf of leaves) {
		const viewInstance = leaf.view as FileExplorerView;

        if(hide) {
            Object.entries(viewInstance.fileItems).forEach(([folderPath, item]) => {
                if(item.file instanceof TFolder) {
                    item.el.toggleClass("import-plugin-hidden",plugin.matchAttachmentFolder(folderPath));
                }
            });
        } else {
            Object.entries(viewInstance.fileItems).forEach(([folderPath, item]) => {
                if(item.file instanceof TFolder) {
                    item.el.toggleClass("import-plugin-hidden",false);
                }
            });
		};
	}
}
```

Can you check what the link the beginning ` at updateVisibilityAttachmentFolders (VM372 plugin:import-attachments-plus:1195:14)` points to? What line `updateVisibilityAttachmentFolders`?

If you want to find out more, I upload a new `main.js` [file](https://github.com/user-attachments/files/17769930/main.js.zip) with `debugger` instructions that stops before and inside the `updateVisibilityAttachmentFolders`. Then you can advance one step at the time and figure out what breaks.

**Comment by ksdavidc** [comment]
I am learning quite a lot from you! Thank you.

> Can you check what the link the beginning at
updateVisibilityAttachmentFolders (VM372
plugin:import-attachments-plus:1195:14) points to? What line
updateVisibilityAttachmentFolders?

The line it stops at is the:

Object.entries(viewInstance.fileItems).forEach(([folderPath, item]) => {

that is inside the if(hide) {



> If you want to find out more, I upload a new main.js file
<https://github.com/user-attachments/files/17769930/main.js.zip> with
debugger instructions that stops before and inside the
updateVisibilityAttachmentFolders. Then you can advance one step at the
time and figure out what breaks.

I replaced main.js with the new one, opened obsidian, with import
attachment installed, but not enabled. Have enabled it and it is stopped at
line 1225, with the debugger activated. Will step through and see what
happens. Whee.

**Comment by alberti42** [comment]
Great. Indeed line 1225 is where the debugger instruction is placed:
![Bildschirmfoto 2024-11-15 um 12 24 26](https://github.com/user-attachments/assets/4f102d74-b6eb-4c0c-92e6-4485ff94654a)

You will notice, you have two times of buttons to advance through the code in steps. "Step over" and "Step into". There is also a button to return out of the present function where you are, which is sometimes convenient. Most of the time, you do not want to call "step into" because you tend to lose focus on wha the code does, since you plunge always deeper into a function and then into the  function called by the previous function.

Now, the problem in our case occurs indeed inside the function `updateVisibilityAttachmentFolders`. So there you may indeed want to `step into`. However, I also placed a `debugger` instruction directly into `updateVisibilityAttachmentFolders`. So, regardless of what you press, the code will stop in `updateVisibilityAttachmentFolders`.

Note that if the console is hidden, then the code runs normally without stopping at the `debugger` point.

**Comment by ksdavidc** [comment]
Yep. I added some more debuggers, and toggled the pause on uncaught & caught exceptions, and that led to this:

<img width="1303" alt="Screenshot 2024-11-15 at 8 41 33 PM" src="https://github.com/user-attachments/assets/055112fd-1439-42f1-8844-0d47a1880294">

Note that I highlighted viewInstance and I don't really know much of how this works, but is it possible it is bugging out due to viewInstance.fileItems? I can't find fileItems anywhere?

**Comment by ksdavidc** [comment]
I can't believe, but I FOUND IT! Yes, it was fileItems. The key piece of information is 
<img width="952" alt="Screenshot 2024-11-15 at 9 07 23 PM" src="https://github.com/user-attachments/assets/e9ea540b-151e-40ab-90d6-6d8974af53b6"> (https://www.reddit.com/r/ObsidianMD/comments/tai8no/i_built_a_plugin_that_displays_wordpage_count_for/)

Your code, seems to assume that the file explorer is active when it installs (or something like that). I was using File Tree, and when I switched to the file explorer, I was able to install without the error, and the installation persisted over reloads.

And the plugin works.

Ironically I don't need it for the purpose I originally intended it, but I am sure it will come in handy. Thank you so much for all your help making the debugger files and walking me through it! WE DID IT!

**Comment by alberti42** [comment]
Nice! Now we know what the problem was. I will test File Tree plugin and add a check to see whether the standard File Explorer plugin is activated. Indeed I assumed people used File Explorerer, but clearly there are already plugins that superceed it and you are using one of those.

I will not path other plugin to reproduce the same features I wanted to have in File Explorer. Mainly, it hides the attachment folders if you choose to do so from the  options. I like to have for each .md an attachment folder containins the associated files. However, I don't like to see all these folders in the File Explorer because it clutters the view.

**Comment by alberti42** [comment]
I tried "File Tree" and my plugin does not crash. Both can coexist. I also tried to disable the factory plugin "Files" that essentially provides the Explorer plugin in Obsidian, and no crash. Even if I restart Obsidian, I don't see any crash. Uhm.. I am not sure how to reproduce the bug.

**Comment by ksdavidc** [comment]
well, in my case, I find I am still getting the error msg, after all. I haven't had a chance to troubleshoot, but will look at it more closely (now that you have taught me how) and let you know.

Another clues is that it is interfering with my ability to drag and drop lines, which is an experimental feature of the outliner plugin. This is also throwing an error, and they may be related. 

Let me try to suss out what is happening better, since it seems to involve many moving parts that I will try to find a decision tree to isolate. 
I would send you my vault but I should try to pare it down first.

**Comment by ksdavidc** [comment]
Attached is a minimal vault. Here is a video showing how to trigger the error. No other plugins involved.
[ImportAttachmentsVault.zip](https://github.com/user-attachments/files/17787359/ImportAttachmentsVault.zip)

https://github.com/user-attachments/assets/515787df-21f7-4f4a-bd8b-ad243d1a9f81

The outliner stuff was a red herring, unrelated.

**Comment by ksdavidc** [comment]
Here is the connection with Outliner.md plugin
[ImportAttachmentsVault with Outliner.zip](https://github.com/user-attachments/files/17787434/ImportAttachmentsVault.with.Outliner.zip)

https://github.com/user-attachments/assets/a7651485-06e6-43d1-b75f-b4f050e84543

**Comment by ksdavidc** [comment]
sorry, text in videos hard to see....let me know if you can't parse it.

**Comment by alberti42** [comment]
Hey David, yes, a bit hard to read, but let me first try to see whether I can read all it.

**Comment by alberti42** [comment]
Yes, the problem with the outliner plugin is very clear. I will try to reproduce it.

**Comment by alberti42** [comment]
Thank you! I am able to reproduce the error from your minimal vault. I will have a look more in detail report back.

**Comment by alberti42** [comment]
I think the problem is related to a recent change in how Obsidian works. Now the different panels (Views in Obsidian parlance) are not loaded until they are opened by the user. This is a good idea as it makes Obsidian more responsive as it does not have to deal with a lot of panels if those are anyhow hidden or not active.

I will have to think how to patch File Explorer when the view of the plugin is not activated by the user yet. In principle I would like to have a callback function that informs my plugin that File Explorer has now been called by the user. I will think about.

**Comment by alberti42** [comment]
It is very interesting. You have 134 plugins but likely you miss just one of the plugins I used in my vault that forces the File Explorer plugin to load even if the user did not select the view. This is why I did not see the error in my vault — at least this is what I believe now.

I realized that the plugin [Obsidian File Explorer Count](https://github.com/ozntel/file-explorer-note-count) has the same problem you reported for my plugin. So, your report caught a bug at least in two plugins. But you do not use this plugin among the many you have.

I will try to confirm my hypothesis by finding what plugin in vault does the job of loading the File Explorer on start. I would be tempted to force loading the view in the background, because anyhow most of the user would like to use the File Explorer.

I am not sure there is a better way. For sure, Obsidian developers would like to avoid it, because that is how they now transitioned to deferred loading of views.

**Comment by alberti42** [comment]
Oh... I found the plugin that if loaded before my plugin and [Obsidian File Explorer Count](https://github.com/ozntel/file-explorer-note-count) saves them: this is [Obsidian Iconize](https://github.com/FlorianWoelki/obsidian-iconize). I am surprised that you do not use it in your collection. But this is the reason why I did not see the bug in my vault.

**Comment by alberti42** [comment]
And this was the issue that people noticed in September about the Iconize plugin: https://github.com/FlorianWoelki/obsidian-iconize/issues/566. It was about the deferred view as I was suspecting. The developer of Iconize was faster in fixing the problem with his/her plugin than me recognizing there was a problem 😅. His/her fix has also fixed my plugin. But of course, I want to have my plugin run without the need to load Iconize.

**Comment by alberti42** [comment]
I have released a new version [1.5.12](https://github.com/alberti42/obsidian-import-attachments-plus/releases/tag/1.5.12) which should fix the problem.

**Comment by ksdavidc** [comment]
waiting on baited breath for it to appear....

**Comment by alberti42** [comment]
I merged two bug fixes into one release: https://github.com/alberti42/obsidian-import-attachments-plus/releases/tag/1.15.13

I cross fingers that no new problems will emerge. Recently, a lot of things have changed in Obsidian under the hood.

**Comment by ksdavidc** [comment]
Excellent. Working  very nicely for me. Thank You!!!

**Comment by alberti42** [comment]
Thanks David for the feedback and great and inquisitive work in debugging the plugin. I now close this issue. If you still have problems with the Outliner plugin (but you said it was a red herring), please open a separate issue.
